### ISSUE TITLE
[VMS] API endpoint and request method for filtered events server action

### ISSUE THREAD

**Issue by PC98** [GTBitsOfGood/gen-soln#296]
* After #293 is completed, let's make an API endpoint and a fetch request method for this server action.
* The API endponit should use `handleGetRequestWithPayloadResponse`. You may have some TS problems with specifying the `convertQueryToServerActionInput` function argument that `handleGetRequestWithPayloadResponse` expects. DM me and we can work together on this.
* Don't forget to make changes to `config.ts`.

**Comment by matthebrown** [comment]
@PC98 I may be asking the wrong question here, but how should we handle the fact that some parameters may not be passed (i.e. someone may just want to filter by date and not causes) in our query conversion method? Additionally, the parameters are currently passed separately (`time=NWEEK&time=WEEK` rather than `times=NWEEK&WEEK`), so we may need to change either the expected name or the way the parameters are passed.

**Comment by PC98** [comment]
I don't think `times=NWEEK&WEEK` would be a correct format, the `&` is usually used to indicate different query parameters. If you have multiple values for the same parameter, you usually would do `time=NWEEK&time=WEEK`, as we do, or `time=NWEEK,WEEK` (the `,` character would get escaped).

But anyway, this is still a good question. If we don't want to change our `fetchRequestWithPayloadResponse` method, here is how I imagine things will work: 

1. The fetch request will be passed an object of type `FilterPageInformation`; the `PaginateWithFilter` fields of `FilterPageInformation` (like `causes`, `cities`, etc.) will come from the initial server-side response, just like how our upcoming events pagination code works.
2. So `fetchRequestWithPayloadResponse` will be passed an object, something like `{ cities: [], causes: ["ARTS"], times: ["TOMORROW", "WEEK"] /* rest * / }, notice how all the `PaginateWithFilter` fields here will be of type array, and if a user has not selected a filter, it would be the empty array.
3. `fetchRequestWithPayloadResponse` calls `querystringify` on this object, and then when we receive this request on our back-end, `req.query` will represent this data as `{ cities: "", causes: "ARTS", times: "TOMORROW,WEEK"  /* rest * /  }`, so basically all the array fields have been converted to strings.
4. Your conversion method will convert `cities: ""` to `cities: []`, `causes: "ARTS"` to `causes: ["ARTS"]`, etc. You might have to use something like `as FilterValue<"cause">[]` to make TS happy.

We can approach this differently but then we would have to change more of our existing code.

**Comment by PC98** [comment]
Also as an FYI, I am simplifying some pagination code as part of this PR: #318 
Do rebase on top of that before working.

**Comment by matthebrown** [comment]
@PC98 My current implementation works (almost) properly and utilizes the `getFilterValuesInQuery` found in `filters.ts`. 
```
const causes = getFilterValuesInQuery(queryRecord, "causes");
```
However, this does present some issues with typescript, since it is expected the type to be either "cause", "time", or "location". Instead, the `queryRecord` contains the fields "causes", "times", and "locations". One option would be to expand the `FilterValue` type to include the plural forms. Another issue is that there is no validation to make sure the parameters are valid. I'm having trouble finding the best way to do this validation. Is there some easy way to check if each of the filters are valid `FilterOption`s? It's a little difficult since our filters are A. read-only (so no reduce or map) and B. the actual value we need to validate against is in the `value` field of each entry. I considered just looping through and adding each value to a new array, but that seems inefficient. Any thoughts?

**Comment by matthebrown** [comment]
We can cast "causes" and others as a `FilterType`, but that just pushes the errors up to `queryType`.

**Comment by PC98** [comment]
For your first question, let's just reconcile `PaginateWithFilter` with our supported `FilterType`s:
```javascript
type SomeName = { [key in FilterType]: FilterValue<key>[] };
interface PaginateWithFilter extends SomeName {
  totalCount: number; // Filtered events page needs to explicitly display total number of results
}
```
After you make that change^ you may have to refactor whatever code uses `PaginateWithFilter`.

> Another issue is that there is no validation to make sure the parameters are valid.

Don't worry about this; our server action `getFilteredEventsCardData` won't filter using invalid values. Let's just make TS happy at compile time and do the best possible thing we can, like ensuring `lat` and `long` are numbers, `cause` is an array, etc.

**Comment by PC98** [comment]
@matthewbrown77 As an FYI, I am adding a `date` field to `FilterPageRequest` as part of #321; I realized we will need a constant date for our time filter computations.

Also it's okay if we don't finish this task in Sprint 4, we can roll it over so don't worry about it too much.
